### PR TITLE
perf: improve instantation speed by dynamic formatter generation

### PR DESCRIPTION
**Pull request by Uzlopak**
As I see that @trim21 is active in this repo, I wanted to suggest a change, which has maybe more impact than #100

This is just a poc. It is not complete regarding the serialization. 

To have some useful performance benchmark, i disabled stacktrace generation.

before:
```sh
uzlopak@Battlestation:~/workspace/fastify-org/fastify-error$ node benchmarks/instantiate.js
instantiate Error x 5,245,780 ops/sec ±0.37% (192 runs sampled)
instantiate FastifyError x 238,523,493 ops/sec ±0.07% (196 runs sampled)
instantiate FastifySpecialError x 20,584,535 ops/sec ±0.54% (194 runs sampled)
```
after:
```sh
uzlopak@Battlestation:~/workspace/fastify-org/fastify-error$ node benchmarks/instantiate.js
instantiate Error x 5,208,812 ops/sec ±0.22% (192 runs sampled)
instantiate FastifyError x 238,148,483 ops/sec ±0.08% (195 runs sampled)
instantiate FastifySpecialError x 144,541,760 ops/sec ±0.12% (194 runs sampled)
```

Obviously it gets slower if type casting is feature complete. But I assume, that it is still faster than calling util.format every time. 
Maybe it is useless, as stacktrace is always the biggest perf bottleneck. But in the case, somebody disables stacktrace on production services (which is i think anyway best practice) it shoiuld have an impact on perf.

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: lib/formatFn.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+'use strict'
+
+const countSpecifiers = require("./countSpecifiers")
+const specifiersRE = /(%[sdifjoOc%])/g
+
+function formatFn(message) {
+  const amountOfPlaceholders = countSpecifiers(message)
+
+  if (amountOfPlaceholders === 0) {
+    return () => message
```
Comment (trim21) [inline lib/formatFn.js:10]: maybe just use string variable in `FastifyError`

Comment (Uzlopak) [reply lib/formatFn.js:10]: More likely to build FastifyError dynamically.

Comment (trim21) [reply lib/formatFn.js:10]: I didn't get it

Comment (Uzlopak) [reply lib/formatFn.js:10]: nevermind. 

Anyhow. this is just a poc. If there is interest, I could improve it.

**Comment by Uzlopak** [comment]
@ivan-tymoshenko 
This is maybe very similar to fast-json-stringify

**Comment by ivan-tymoshenko** [comment]
@Uzlopak sorry, I don't understand. Can you explain, please?

**Comment by Uzlopak** [comment]
new benchmarks:
before:
node benchmarks/instantiate.js
instantiate Error x 5,148,366 ops/sec ±0.28% (191 runs sampled)
instantiate FastifyError x 121,849,157 ops/sec ±0.50% (190 runs sampled)
instantiate FastifyError arg 1 x 14,033,361 ops/sec ±0.25% (194 runs sampled)
instantiate FastifyError arg 2 x 10,830,852 ops/sec ±0.33% (191 runs sampled)

after:
node benchmarks/instantiate.js
instantiate Error x 5,049,694 ops/sec ±0.26% (193 runs sampled)
instantiate FastifyError x 236,957,612 ops/sec ±0.09% (193 runs sampled)
instantiate FastifyError arg 1 x 83,758,765 ops/sec ±0.91% (185 runs sampled)
instantiate FastifyError arg 2 x 63,728,097 ops/sec ±2.10% (179 runs sampled)

**Comment by Uzlopak** [comment]
@ivan-tymoshenko 

I use now the message string as pattern to create the format function once instead of calling format(message, args) every time.

**Comment by Uzlopak** [comment]
With default stackTraceLimit

before:
node benchmarks/instantiate.js
instantiate Error x 322,998 ops/sec ±2.65% (137 runs sampled)
instantiate FastifyError x 231,124 ops/sec ±1.33% (188 runs sampled)
instantiate FastifyError arg 1 x 171,481 ops/sec ±1.44% (182 runs sampled)

after:
node benchmarks/instantiate.js
instantiate Error x 304,488 ops/sec ±3.00% (138 runs sampled)
instantiate FastifyError x 224,038 ops/sec ±1.68% (187 runs sampled)
instantiate FastifyError arg 1 x 188,652 ops/sec ±1.69% (188 runs sampled)
instantiate FastifyError arg 2 x 180,878 ops/sec ±1.80% (176 runs sampled)

In conclusio: 

It is signigifcant faster if we set stackTraceLimit to 0. If we use default stackTraceLimit, we are not definetly not faster.
instantiate FastifyError arg 2 x 188,166 ops/sec ±1.61% (181 runs sampled)

File: benchmarks/instantiate.js [hunk 2]
```diff
@@ -3,14 +3,19 @@
 const benchmark = require('benchmark')
 const createError = require('..')
 
+Error.stackTraceLimit = 0
```
Comment (trim21) [inline benchmarks/instantiate.js:6]: there is already `benchmarks/no-stack.js` for this?

**Review by mcollina — CHANGES_REQUESTED** [review]
I don't think benchmarking without stacktraces is worthwhile. Stacktraces are incredibly useful to determine where a problem originated.

What is the goal of the change? What are are you trying to achieve?

**Comment by jsumners** [comment]
I think we already had a discussion and determined that we do not want to omit stack traces. They are the purpose of the error.

**Comment by trim21** [comment]
> What is the goal of the change? What are are you trying to achieve?

improve string formatting perf. disabling stack trace is would make it easier to focus on the perf of string formatting, not actually disable it for perf.

**Comment by Uzlopak** [comment]
I think, that in production Environment disabling stacktraces is a viable decision. In development using stacktraces make obviously sense.

**Comment by jsumners** [comment]
Sure, it's a viable decision. But I would recommend against it. Without the stack trace you are left with very little information to determine the cause of issues.

**Pull request closed without merging**